docs/design/server42/README.md

```diff
@@ -0,0 +1,235 @@
+  Kafka message, and sends it to the Stats Topology through the Kafka queue.
+  It must support more than one copy for scaling.
+- Stats storm topology is written in Java. It gets a big data packet from Stats
+  parses it, and puts the information to OpenTSDB.
+- Control application is written in Java. It a control plane application
+  for adding/removing flow on Server 42 and for providing other control features.
+- Control storm topology is written in Java. It gets a request from NB
+  and FROM other topologies. Also, it adds new OpenFlow rules to switches through 
+  Floodlight and reads the list of the Server 42 instances from the config file.
+  
+#### DPDK setup
+- <https://doc.dpdk.org/guides/linux_gsg/linux_drivers.html>
+- <https://pktgen-dpdk.readthedocs.io/en/latest/getting_started.html>
+#### DPDK core binding
+VM with Server 42 must be bound to minimum 4 dedicated hardware CPUs in one NUMA group.
+### Communication
+All communication between Control/Stat applications and storm topologies is operated
+by Kafka with Google protocol buffers. The communication with Server 42
+is handled by ZeroMQ as transport and Google protocol buffers.
+### Seq diagram
+#### Control Plane
+![Control Plane](cp.png)
+#### Packet and Stats processing
+![Packet and Stats processing](packet_process.png)
+### API
+#### Server 42 API
+
+Based on [Google Protobuf](https://developers.google.com/protocol-buffers/docs/proto3)
+
+##### Flow
+
+###### Common Types
+```
+message Flow {
+  string flow_id = 1;
+    enum EncapsulationType {
+        VLAN = 0;
+        VXVLAN = 1;
+      }
+  EncapsulationType encapsulation_type = 2;
+  int64 tunnel_id = 3;
+}
+```
+
+##### Command packet
+```
+message CommandPacket {
+    enum Type {
+        ADD_FLOW = 0;
+        REMOVE_FLOW = 1;
+        CLEAR_FLOWS = 2;
+        LIST_FLOWS = 3;
+        PUSH_SETTINGS = 4;
+      }
+  Type type = 1;
+  optional google.protobuf.Any command = 2;
+}
+```
+[google.protobuf.Any](https://developers.google.com/protocol-buffers/docs/proto3#any)
+
+The command must be one of the following:
+
+###### Add Flow
+```
+message AddFlow {
+  Flow flow = 1;
+}
+```
+###### Remove Flow
+```
+message RemoveFlow {
+  Flow flow = 1;
+}
+```
+###### Clear Flows
+
+Payload not allowed here.
+
+###### List Flows
+```
+message ListFlows {
+  repeated Flow flow = 1;
+}
+```
+###### Push settings
+
+```
+message PushSettings {
+  int32 packet_generation_interval_in_ms = 1;
+}
+```
+
+#### Stats API
+```
+message FlowLatencyPacket {
+    bytes flow_id = 1;
+    int64 t0 = 2;
+    int64 t1 = 3;
+    uint64 packet_id = 4;
+}
+
+message FlowLatencyPacketBucket {
+    repeated FlowLatencyPacket packet = 1;
+}
+
+```
+
+#### NB API
+- enable/disable all flow
+- enable/disable all flow per switch/server42 instances/POP
+- get list of enabled flows
+- enable/disable flow
+- push settings
+- list server42 instances(POP)
+- add/remove server42 instances(POP)
```

Comment:
Could you provide more details on how a server is added? Will it be a new VM?

Reply:
that is related to concrete customer infrastructure. for common, this is a couple of VMs.